internal/controller/controllers/agent_controller.go

```diff
@@ -113,6 +155,64 @@ func (r *AgentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
 	return r.updateStatus(ctx, agent, host, nil, false)
 }
 
+func (r *AgentReconciler) deleteAgent(ctx context.Context, agent types.NamespacedName) (ctrl.Result, error) {
+	agentToDelete := &aiv1beta1.Agent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      agent.Name,
+			Namespace: agent.Namespace,
+		},
+	}
+	if delErr := r.Client.Delete(ctx, agentToDelete); delErr != nil {
+		r.Log.WithError(delErr).Errorf("Failed to get delete resource %s %s", agent.Name, agent.Namespace)
+		return ctrl.Result{Requeue: true}, delErr
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *AgentReconciler) deregisterHostIfNeeded(ctx context.Context, key types.NamespacedName) (ctrl.Result, error) {
+
+	buildReply := func(err error) (ctrl.Result, error) {
+		reply := ctrl.Result{}
+		if err == nil {
+			return reply, nil
+		}
+		reply.RequeueAfter = defaultRequeueAfterOnError
+		err = errors.Wrapf(err, "failed to deregister host: %s", key.Name)
+		r.Log.Error(err)
+		return reply, err
+	}
+
+	h, err := r.Installer.GetHostById(key.Name) // TODO: Change implementation to GetHostByKubeKey after MGMT-6006
+
+	if errors.Is(err, gorm.ErrRecordNotFound) {
+		// return if from any reason host is already deleted from db (or never existed)
+		return buildReply(nil)
+	}
+
+	if err != nil {
```

Comment:
NIT: `if` -> `else if `

Reply:
I reordered it so we first check if `err != nil` and drill down to ask specifically about `gorm.ErrRecordNotFound` inside that section